server/app/services/applicant/exception/DuplicateApplicationException.java

```diff
@@ -0,0 +1,11 @@
+package services.applicant.exception;
+
+/**
+ * Thrown if an application is submitted that is a duplicate of a previously submitted one. Triggers
+ * a screen to be shown to the applicant asking them to continue editing or exit the application.
+ */
+public class DuplicateApplicationException extends RuntimeException {
```

Comment:
I'm not very familiar with the best practices regarding Exception handling, so this may be off base.

Would it be safer to make this a checked exception? Having an Exception inadvertently thrown by a controller will result an error in the browser, which doesn't make anyone happy.

Reply:
Thinking about _Effective Java_ items 69 and 70, I'm not sure we should be using exceptions here at all.

- 69 says that "exceptions should be exceptional" and checking if this is a duplicate feels more like a "normal control flow" sort of thing. I could be convinced that this is no less exceptional then "program not found", which we do use an exception for, but at the moment a 404 seems more exceptional then submitting a duplicate program
- 70 says that we should used checked exceptions for recoverable conditions and runtime exceptions for programming errors

Links (I couldn't find a great, linkable version of _Effective Java_ so this will have to do):
- https://github.com/AlphaWang/alpha-effective-java-3e/blob/master/10_exceptions/item_69_use_exceptions_only_for_exceptional_conditions.md#item-69-use-exceptions-only-for-exceptional-conditions
- https://github.com/AlphaWang/alpha-effective-java-3e/blob/master/10_exceptions/item_70_use_checked_exceptions_for_recoverable_conditions_and_runtime_exceptions_for_programming_errors.md

Reply:
I never much cared for the `ProgramNotFoundException` or `QuestionNotFoundException`. Something like a Null Object or other pattern could have probably taken the place for those and provided more flexibility. I'd recommend we don't continue to propagate this pattern, but if we have infrastructure components already build to use that pattern I'd go with it in this case and have a larger refactor issue made.

Also checked exception are the devil and no one will convince me otherwise.

Reply:
I agree that this is an unusual use of exceptions, but, in this case, I don't see how it will cause issues other than being annoying.  I wanted to suggest that we refactor by having the `ApplicationRepository.submitApplicationInternal()` return a null object, but then we'd have no way of knowing why the method returned null, assuming there can be reasons other than a duplicate application.

Let's definitely make it a checked exception, though.

Reply:
I think the refactor would look like creating an `ApplicationSubmissionResult` object that contains a field for the submitted application and a field for any errors, and returning that up the tree instead of the `Application` that we currently do. Similarly, methods like `validateApplicationForSubmission()` would need to return something like a `ApplicationValidationResult`.

Given that it would probably be more confusing to have some errors bubble up with the current exception strategy and others with this wrapper object, I'm ok with leaving it as you have it in the PR now.

Reply:
@avaleske I like that idea for the refactor.  My vote is that we create a separate tech debt issue to refactor all of these un-exceptional exceptions.  :-)

Reply:
I like that idea for a refactor a lot @avaleske! I'll follow @melanie-exygy suggestion and create an issue to do that before closing out this work.

Reply:
I added this issue: https://github.com/civiform/civiform/issues/5775

Feel free to comment/add to it.